onnxruntime/core/providers/cpu/tensor/transpose.cc

```diff
@@ -15,23 +15,65 @@ namespace onnxruntime {
    etc.
    */
 
-// ComputeOffset: compute offset into a tensor. This is essentially the dot-product of
-// index and stride, restricted to the specified number of axes.
-static inline size_t ComputeOffset(const std::vector<int64_t>& index, const std::vector<size_t>& stride, int64_t num_axes) {
-  size_t offset = 0;
-  for (int64_t j = 0; j < num_axes; ++j) {
-    offset += index[j] * stride[j];
+typedef struct MultiIndex {
+  size_t index;
+  size_t upper_bound;
+  int64_t stride;
+  MultiIndex() {}
+  MultiIndex(size_t i, size_t n, int64_t s) {
+    index = i;
+    upper_bound = n;
+    stride = s;
+  }
+} MultiIndex;
+
+static size_t IncrementIndexAndComputeOffsetSetup(MultiIndex* mindex, int64_t num_axes, const std::vector<int64_t>& target_dims,
+                                                  const std::vector<size_t>& stride, size_t element_size) {
+  size_t naxes = 0;
+  for (int64_t i = 0; i < num_axes; ++i) {
+    if (target_dims[i] == 1)
+      continue;
+    mindex[naxes] = MultiIndex(0, static_cast<size_t>(target_dims[i]), stride[i] * element_size);
+    ++naxes;
+  }
+  return naxes;
+}
+
+// Combines multi-index increment and corresponding pointer in the tensor to transpose.
+static void IncrementIndexAndComputeOffset(MultiIndex* mindex, size_t naxes, const uint8_t*& local_source) {
+  MultiIndex* it = mindex + (naxes - 1);
+  local_source += it->stride;
+  if (++it->index < it->upper_bound)
+    return;
+  local_source -= it->stride * it->index;
+  it->index = 0;
+  --it;
+  MultiIndex* rend = mindex - 1;
+  for (; it != rend; --it) {
+    local_source += it->stride;
+    if (++it->index < it->upper_bound)
+      break;
+    local_source -= it->stride * it->index;
+    it->index = 0;
   }
-  return offset;
 }
 
-// IncrementIndex: Increment an index into a tensor (in lexicographic ordering), wrapping
-// around the specified upper_bound.
-static inline void IncrementIndex(std::vector<int64_t>& index, const std::vector<int64_t>& upper_bound, int64_t num_axes) {
-  for (int64_t k = num_axes - 1; k >= 0; --k) {
-    index[k]++;
-    if (index[k] < upper_bound[k]) break;
-    index[k] = 0;
+// Combines multi-index increment and corresponding pointer in the string tensor to transpose.
+static void IncrementIndexAndComputeOffset(MultiIndex* mindex, size_t naxes, const std::string*& local_source) {
+  MultiIndex* it = mindex + (naxes - 1);
```

Comment:
naxes is size_t type. naxes - 1 is dangerous unless you explicitly say naxes won't be zero.

Reply:
I added parenthesis to get ``(mindex + naxes) - 1``, this should solve this issue.